modules/grpc/bigquery/bigquery-worker.cpp

```diff
@@ -0,0 +1,516 @@
+  finalize_request.set_name(write_stream.name());
+
+  status = this->stub->FinalizeWriteStream(&ctx, finalize_request, &finalize_response);
+  if (!status.ok())
+    {
+      msg_warning("Error finalizing BigQuery write stream", evt_tag_str("error", status.error_message().c_str()),
+                  evt_tag_str("details", status.error_details().c_str()),
+                  log_pipe_location_tag((LogPipe *) this->super->super.owner));
+    }
+
+  this->connected = false;
+}
+
+void
+DestinationWorker::prepare_batch()
+{
+  this->batch_size = 0;
+  this->current_batch = google::cloud::bigquery::storage::v1::AppendRowsRequest{};
+
+  this->current_batch.set_write_stream(write_stream.name());
+  this->current_batch.set_trace_id("syslog-ng-bigquery");
+  google::cloud::bigquery::storage::v1::AppendRowsRequest_ProtoData *proto_rows =
+    this->current_batch.mutable_proto_rows();
+  google::cloud::bigquery::storage::v1::ProtoSchema *schema = proto_rows->mutable_writer_schema();
+  this->get_owner()->schema_descriptor->CopyTo(schema->mutable_proto_descriptor());
+}
+
+void
+DestinationWorker::format_template(LogTemplate *tmpl, LogMessage *msg, GString *value, LogMessageValueType *type)
+{
+  DestinationDriver *owner = this->get_owner();
+
+  if (log_template_is_trivial(tmpl))
+    {
+      gssize trivial_value_len;
+      const gchar *trivial_value = log_template_get_trivial_value_and_type(tmpl, msg, &trivial_value_len, type);
+
+      g_string_truncate(value, 0);
+      if (trivial_value)
+        g_string_append_len(value, trivial_value, trivial_value_len);
+
+      return;
+    }
+
+  LogTemplateEvalOptions options = {&owner->template_options, LTZ_SEND, this->super->super.seq_num, NULL, LM_VT_STRING};
+  log_template_format_value_and_type(tmpl, msg, &options, value, type);
+}
+
+bool
+DestinationWorker::insert_field(const google::protobuf::Reflection *reflection, const Field &field,
+                                LogMessage *msg, google::protobuf::Message *message)
+{
+  DestinationDriver *owner = this->get_owner();
+
+  ScratchBuffersMarker m;
+  GString *value = scratch_buffers_alloc_and_mark(&m);
+
+  LogMessageValueType type;
+
+  this->format_template(field.value, msg, value, &type);
+
+  if (type == LM_VT_NULL)
+    {
+      if (field.field_desc->is_required())
+        {
+          msg_error("Missing required field", evt_tag_str("field", field.name.c_str()));
+          goto error;
+        }
+
+      scratch_buffers_reclaim_marked(m);
+      return true;
+    }
+
+  switch (field.field_desc->cpp_type())
+    {
+    case FieldDescriptor::CppType::CPPTYPE_STRING:
+      reflection->SetString(message, field.field_desc, std::string{value->str, value->len});
+      break;
+    case FieldDescriptor::CppType::CPPTYPE_INT32:
+    {
+      int32_t v;
+      if (!type_cast_to_int32(value->str, &v, NULL))
+        {
+          type_cast_drop_helper(owner->template_options.on_error, value->str, "integer");
```

Comment:
The `on-error()`'s value is not considered in this function, we always drop the whole message even if `drop-property` is set. IMO we should either handle all the possible `on-error()` cases as expected or should fail the init if a not supported `on-error()` is set by the user.

Reply:
We don't drop anything here, I'm just using the helper, but not making the actual drop. The drop happens outside.

Reply:
(drop-property should work, but non-nullable fields will cause an error in all cases, as BigQuery follows traditional SQL schemas)

Reply:
You are right, the `ON_ERROR_DROP_MESSAGE` is checked manually in `DestinationWorker::insert`. I was just used to see the `type_cast_drop_helper()` decide whether to drop or not.

The `on-error("fallback-to-string")` case still does not do what it says, but in this case I think we cannot support it, as the columns have a strict type. `on-error("fallback-to-string")` is an invalid config, we could warn the user, refuse to start or do nothing. I would prefer to do something, but I think there are other drivers/cases where we do nothing, so I do not push it.